Library/OcVariableRuntimeLib/OcVariableRuntimeLib.c

```diff
@@ -0,0 +1,621 @@
+    // present, we always write non-volatile. (This issue does not affect saving NVRAM
+    // from OS since it saves everything in its filter without checking for V/NV.)
+    //
+    for (VariableIndex = 0; VariableIndex < VariableMap->Count; ++VariableIndex) {
+      OcDirectSetNvramVariable (
+        OC_BLOB_GET (VariableMap->Keys[VariableIndex]),
+        &VariableGuid,
+        OPEN_CORE_NVRAM_NV_ATTR, // mNvramConfig->WriteFlash ? OPEN_CORE_NVRAM_NV_ATTR : OPEN_CORE_NVRAM_ATTR,
+        VariableMap->Values[VariableIndex]->Size,
+        OC_BLOB_GET (VariableMap->Values[VariableIndex]),
+        SchemaEntry,
+        mNvramConfig->LegacyOverwrite
+        );
+    }
+  }
+
+  OC_NVRAM_STORAGE_DESTRUCT (&NvramStorage, sizeof (NvramStorage));
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+DeleteFile (
+  IN EFI_FILE_PROTOCOL  *Directory,
+  IN CONST CHAR16       *FileName
+  )
+{
+  EFI_STATUS  Status;
+
+  Status = OcDeleteFile (Directory, FileName);
+  if (EFI_ERROR (Status)) {
+    if (Status == EFI_NOT_FOUND) {
+      Status = EFI_SUCCESS;
+    }
+  }
+
+  return Status;
+}
+
+//
+// Serialize one section at a time, NVRAM scan per section.
+//
+STATIC
+OC_PROCESS_VARIABLE_RESULT
+EFIAPI
+SerializeSectionVariables (
+  IN EFI_GUID  *Guid,
+  IN CHAR16    *Name,
+  IN VOID      *Context
+  )
+{
+  EFI_STATUS          Status;
+  NVRAM_SAVE_CONTEXT  *SaveContext;
+  UINT32              Attributes;
+  UINTN               DataSize;
+  UINTN               Base64Size;
+  UINTN               Base64Pos;
+
+  ASSERT (Context != NULL);
+  SaveContext = Context;
+
+  if (!CompareGuid (Guid, &SaveContext->SectionGuid)) {
+    return OcProcessVariableContinue;
+  }
+
+  if (!OcVariableIsAllowedBySchemaEntry (SaveContext->SchemaEntry, Guid, Name, OcStringFormatUnicode)) {
+    return OcProcessVariableContinue;
+  }
+
+  do {
+    DataSize = SaveContext->DataBufferSize;
+    Status   = gRT->GetVariable (
+                      Name,
+                      Guid,
+                      &Attributes,
+                      &DataSize,
+                      SaveContext->DataBuffer
+                      );
+    if (Status == EFI_BUFFER_TOO_SMALL) {
+      while (DataSize > SaveContext->DataBufferSize) {
+        if (OcOverflowMulUN (SaveContext->DataBufferSize, 2, &SaveContext->DataBufferSize)) {
+          SaveContext->Status = EFI_OUT_OF_RESOURCES;
+          return OcProcessVariableAbort;
+        }
+      }
+
+      FreePool (SaveContext->DataBuffer);
+      SaveContext->DataBuffer = AllocatePool (SaveContext->DataBufferSize);
+      if (SaveContext->DataBuffer == NULL) {
+        SaveContext->Status = EFI_OUT_OF_RESOURCES;
+        return OcProcessVariableAbort;
+      }
+    }
+  } while (Status == EFI_BUFFER_TOO_SMALL);
+
+  if (EFI_ERROR (Status)) {
+    SaveContext->Status = Status;
+    return OcProcessVariableAbort;
+  }
+
+  //
+  // Only save non-volatile variables; also, match launchd script and only save
```

Comment:
Ugh, cannot we use OcXmlLib there?

Reply:
You mean as opposed to me currently manually building up the XML 'by hand' in a string buffer?

If so, then the answer is that currently OcXmlLib has no write support. (Even if it was considered preferable to add that, that is another fairly major project...! :-// )

Reply:
Can we postpone this one for now, too?

Reply:
Ok.